src/shipper.py

```diff
@@ -26,34 +32,61 @@ class UnknownURL(Exception):
 
 
 class LogzioShipper(object):
-    MAX_BULK_SIZE_IN_BYTES = 1 * 1024 * 1024
+    MAX_BULK_SIZE_IN_BYTES = 3 * 1024 * 1024
 
     def __init__(self, logzio_url):
-        self._size = 0
-        self._logs = []
+        self._decompress_size = 0
+        self._logs = StringIO.StringIO()
         self._logzio_url = logzio_url
+        self._logs_counter = 0
+        try:
+            self._compress = os.environ['COMPRESS'].lower() in ["yes", "true", "t", "1"]
+        except KeyError:
+            self._compress = False
+        self._writer = self._create_new_writer()
+
+    def _create_new_writer(self):
+        self._logs.truncate(0)
+        return gzip.GzipFile(mode='wb', fileobj=self._logs) if self._compress else self._logs
 
     def add(self, log):
         # type: (dict) -> None
         json_log = json.dumps(log)
-        self._logs.append(json_log)
-        self._size += sys.getsizeof(json_log)
-        self._try_to_send()
+        self._writer.write("\n" + json_log) if self._decompress_size else self._writer.write(json_log)
+        self._logs_counter += 1
+        self._decompress_size += sys.getsizeof(json_log)
+        # To prevent flush() after every log added
+        if self._decompress_size > self.MAX_BULK_SIZE_IN_BYTES:
```

Comment:
if you wrap the writer then you can do a "if writer.logsize() > max bulk size" thing and it will be cleaner